### PR TITLE
Increasing cross-scheme API consistency and implementation

### PR DESCRIPTION
**Pull request by ChristianMct**
This PR regroups several work items related to software engineering and API design:

- [x] providing a simpler key format for rotation keys and simpler rotation evaluation logic
- [x] providing a common interface and implementation for the generation of keys in `dbfv` and `dckks`
- [x] harmonizing the API and struct names in `bfv` and `ckks`
- [x] use of the same underlying type for keys in `bfv` and `ckks`
- [x] changing the `evaluators`s API to better support parallel evaluation

### CODE REVIEW

**Comment by Pro7ech** [comment]
- [ ] Evaluator should check that the keys are available 
- [ ] Now that the keys are in the evaluator, the API should offer the possibility to do multiplications with or without relinearization (this was previously done by passing a nil pointer)

**Review by ChristianMct — COMMENTED** [review]
Pass on the full PR done.

**Review by Pro7ech — APPROVED** [review]

File: bfv/keygen.go [hunk 1]
```diff
@@ -219,207 +140,72 @@ func (keygen *keyGenerator) GenRelinKey(sk *SecretKey, maxDegree uint64) (evk *E
...
-// GenRotationKeysPow2 generates a new rotation key with all the power-of-two rotations to the left and right, as well as the conjugation.
-func (keygen *keyGenerator) GenRotationKeysPow2(skOutput *SecretKey) (rotKey *RotationKeys) {
-
-	if keygen.ringQP == nil {
-		panic("Cannot GenRotationKeysPow2: modulus P is empty")
+// GenRotationKeysForRotations generates a RotationKeySet supporting left rotations by k positions for all k in ks.
+// Negative k is equivalent to a right rotation by k positions
+// If includeConjugate is true, the resulting set contains the conjugation key.
+func (keygen *keyGenerator) GenRotationKeysForRotations(ks []int, includeConjugate bool, sk *SecretKey) (rks *RotationKeySet) {
+	galEls := make([]uint64, len(ks), len(ks)+1)
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply bfv/keygen.go:178]: It is not needed but avoids the slice extension mechanism to re-allocate the backing array with twice the size (and to copy the content of galEls in the new backing array). Indeed, this is a very minor optimization in this context.

File: bfv/keys.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+func NewSwitchingKey(params *Parameters) *SwitchingKey {
+	return &SwitchingKey{*rlwe.NewSwitchingKey(params.N(), params.QPiCount(), params.Beta())}
+}
+
+// NewRelinearizationKey returns an allocated BFV public relinearization key with zero value for each degree in [2 < maxRelinDegree].
+func NewRelinearizationKey(params *Parameters, maxRelinDegree uint64) *RelinearizationKey {
+	return &RelinearizationKey{*rlwe.NewRelinKey(maxRelinDegree, params.N(), params.QPiCount(), params.Beta())}
+}
+
+// NewRotationKeySet return an allocated set of BFV public relineariation keys with zero values for each galois element
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply bfv/keys.go]: Done.

File: bfv/bfv_test.go [hunk 3]
```diff
@@ -116,6 +116,16 @@ func testParameters(testctx *testContext, t *testing.T) {
 		assert.NoError(t, err)
 		assert.True(t, p.Equals(testctx.params))
 	})
+
+	t.Run("Parameters/InverseGaloisElement/", func(t *testing.T) {
+		for i := 1; i < int(testctx.params.N()/2); i++ {
+			galEl := testctx.params.GaloisElementForColumnRotationBy(i)
+			mod := 2 * testctx.params.N()
+			inv := testctx.params.InverseGaloisElement(galEl)
+			res := (inv * galEl) % mod
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply bfv/bfv_test.go:125]: When there is no need for performance, I'd stick to the most straightforward/explicit expression.

**Pull request merged**